src/PrizeVault.sol

```diff
@@ -0,0 +1,750 @@
+     * @param _r R part of the secp256k1 signature
+     * @param _s S part of the secp256k1 signature
+     * @return Amount of Vault shares minted to `_owner`.
+     */
+    function depositWithPermit(
+        uint256 _assets,
+        address _owner,
+        uint256 _deadline,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+    ) external returns (uint256) {
+        if (_owner != msg.sender) {
+            revert PermitCallerNotOwner(msg.sender, _owner);
+        }
+
+        IERC20Permit(address(_asset)).permit(_owner, address(this), _assets, _deadline, _v, _r, _s);
+
+        uint256 _allowance = _asset.allowance(_owner, address(this));
+        if (_allowance != _assets) {
+            revert PermitAllowanceNotSet(_owner, address(this), _assets, _allowance);
+        }
+
+        uint256 _shares = previewDeposit(_assets);
+        _depositAndMint(_owner, _owner, _assets, _shares);
+        return _shares;
+    }
+
+    /**
+     * @notice Deposit assets into the Vault and delegate to the sponsorship address.
+     * @param _assets Amount of assets to deposit
+     * @return Amount of shares minted to caller.
+     */
+    function sponsor(uint256 _assets) external returns (uint256) {
+        address _owner = msg.sender;
+
+        uint256 _shares = previewDeposit(_assets);
+        _depositAndMint(_owner, _owner, _assets, _shares);
+
+        if (twabController.delegateOf(address(this), _owner) != SPONSORSHIP_ADDRESS) {
+            twabController.sponsor(_owner);
+        }
+
+        emit Sponsor(_owner, _assets, _shares);
+
+        return _shares;
+    }
+
+    /* ============ Yield Functions ============ */
+
+    /**
+     * @notice Total available yield on the vault.
+     * @return The available yield balance
+     */
+    function availableYieldBalance() public view returns (uint256) {
+        uint256 _totalAssets = totalAssets();
+        uint256 _allocatedAssets = totalSupply() + yieldBuffer;
+        if (_allocatedAssets >= _totalAssets) {
+            return 0;
+        } else {
+            unchecked {
+                return _totalAssets - _allocatedAssets;
+            }
+        }
+    }
+
+    /* ============ LiquidationSource Functions ============ */
+
+    /// @inheritdoc ILiquidationSource
+    /// @dev Returns the withdrawable amount of the available yield minus any yield fees.
+    function liquidatableBalanceOf(address _token) public view returns (uint256) {
+        if (_token != address(_asset)) {
+            return 0;
+        } else {
+            uint256 _availableYieldBalance = availableYieldBalance();
+            uint256 _maxWithdraw = yieldVault.maxWithdraw(address(this)) + _asset.balanceOf(address(this));
+            uint256 _withdrawableYieldBalance = _availableYieldBalance >= _maxWithdraw ? _maxWithdraw : _availableYieldBalance;
+            return _withdrawableYieldBalance.mulDiv(FEE_PRECISION - yieldFeePercentage, FEE_PRECISION);
+        }
```

Comment:
This function is only considering liquid assets in the underlying yield source, instead of considering all of the assets (including debt).

This means that if, in the case of Aave, 1000 USDC has accrued as yield but only 100 USDC are liquid, then the vault can only liquidate 100 USDC.

The vault should be able to liquidate debt as well!

Reply:
I see your point, but it still seems a little counter-intuitive since a bot would be liquidating non-liquid debt tokens and would then have to rely on debt liquidity in a pool like uniswap to complete the cycle. If we are trying to move away from prize vault share liquidity with the new deployment, then we wouldn't want to rely on LPs for liquidations since it's using WETH as the prize asset and we don't want WETH being won and burned by LPs.

I think we should weigh the pros and cons of both and evaluate it from multiple perspectives: gas-cost, effects of share LPs, and bot convenience. I'm open to switching it back to share minting, but I want to make sure we're not missing out on anything.

Reply:
As an example for why withdrawing assets can be beneficial we can consider what flash swap liquidators do:

1. get prize vault asset (example pUSDC)
2. swap vault asset from pUSDC -> USDC -> ... -> prizeAsset
3. contribute prizeAsset to prize pool and keep profit

This falls apart when there is no (pUSDC -> USDC) liquidity, which is the reason why we don't currently have flash swaps for any vaults other than pUSDC.e and the original pWETH. In addition, if the pUSDC -> USDC LP becomes unbalanced, then the liquidation might have a worse exchange rate.

To fix this in the flash swap implementation, the flash swap contract would have to do the following:

1. get pUSDC from vault (which mints the tokens)
2. withdraw USDC (which burns the tokens and withdraws the assets)
3. swap from USDC -> ... -> prizeAsset
4. contribute prizeAsset to prize pool

The minting of shares becomes gas waste in this example since they are being immediately burned in the following step.

So I think it comes down to if we expect every vault to have prize vault share liquidity or not.

Reply:
Yes; let's go with minting shares. It creates optionality; the liquidator can decide to consume only the liquid yield or take the full yield including debt. Removing that optionality means that everyone is limited to only the liquid tokens in the underlying yield source; as we saw with the previous version of the vault that was problematic!

Reply:
How do you feel about supporting both so the best strategy can be chosen by the vault deployer based on the restrictions of the yield vault? The following commit will let the liquidation pair do either strategy: https://github.com/GenerationSoftware/pt-v5-vault/pull/75/commits/9386ea1c2d389b8a3a717967a76e8a752efb7ecb

That way other fully liquid yield vaults (not aave) can let bots liquidate assets if gas costs are favorable.